File: java/src/array/SinglyNoTailLinkedListStack.java
package array;

public class SinglyNoTailLinkedListStack<E> {

    private Node head;

    private int size;

    private class Node {
        private E key;
        private Node next;

        public Node(E key) {
            this.key = key;
        }
    }

    public String toString() {
        Node tmp = head;

        StringBuffer sb = new StringBuffer();
        sb.append("[");
        for (int i = 0 ; i < size; i++) {
            sb.append(tmp.key);
            sb.append(",");
            tmp = tmp.next;
        }
        sb.deleteCharAt(sb.length() - 1);
        sb.append("]");

        return sb.toString();
    }

    public void push(E input) {
        Node node = new Node(input);

        node.next = head;
        head = node;
        size++;
    }

    public E top() {
        if (head == null)
            return null;
        return head.key;
    }

    public E pop() {
        if (head == null)
            return null;

        Node del = head;

        head = del.next;
        E returnValue = del.key;
        del = null;
        size--;

        return returnValue;
    }

    public boolean empty() {
        return head == null;
    }
}

File: python/array/singlyNoTailLinkedListStack.py
class Stack:
    head = None
    size = 0

    class Node:
        next = None
        def __init__(self, data):
            self.data = data

    def toString(self):
        tmp = self.head
        result = []
        for i in range(self.size):
            result.append(str(tmp.data))
            tmp = tmp.next
        return ','.join(result)

    def push(self, input):
        node = self.Node(input)
        node.next = self.head
        self.head = node
        self.size += 1

    def top(self):
        if self.head == None:
            return None
        else:
            return self.head.data
    
    def pop(self):
        if self.head == None:
            return None
        node = self.head
        self.head = node.next
        result = node.data
        node = None
        self.size -= 1
        return result
    
    def empty(self):
        return self.head == None
